examples/PingPong.hs

```diff
@@ -37,38 +45,79 @@ instance Mockable SharedAtomic IO where
     liftMockable (NewSharedAtomic t) = Conc.newMVar t
     liftMockable (ModifySharedAtomic atomic f) = Conc.modifyMVar atomic f
 
-type instance ChannelT IO = Conc.Chan
+type instance ChannelT IO = Conc.TChan
 
 instance Mockable Channel IO where
-    liftMockable (NewChannel) = Conc.newChan
-    liftMockable (ReadChannel channel) = Conc.readChan channel
-    liftMockable (WriteChannel channel t) = Conc.writeChan channel t
+    liftMockable (NewChannel) = STM.atomically Conc.newTChan
+    liftMockable (ReadChannel channel) = STM.atomically $ Conc.readTChan channel
+    liftMockable (TryReadChannel channel) = STM.atomically $ Conc.tryReadTChan channel
+    liftMockable (UnGetChannel channel t) = STM.atomically $ Conc.unGetTChan channel t
+    liftMockable (WriteChannel channel t) = STM.atomically $ Conc.writeTChan channel t
 
 instance Mockable Bracket IO where
     liftMockable (Bracket acquire release act) = Exception.bracket acquire release act
 
 instance Mockable Throw IO where
     liftMockable (Throw e) = Exception.throwIO e
 
-workers :: ( RandomGen g ) => NodeId -> g -> [NodeId] -> [Worker IO]
+instance Mockable Catch IO where
+    liftMockable (Catch action handler) = action `Exception.catch` handler
+
+-- Sending a message which encodes to "" is problematic!
+-- The receiver can't distinuish this from the case in which the sender sent
+-- nothing at all.
+-- So we give custom Ping and Pong types with non-generic Binary instances.
+--
+-- TBD should we fix this in network-transport? Maybe every chunk is prefixed
+-- by a byte giving its length? Wasteful I guess but maybe not a problem.
```

Comment:
@dcoutts This annoys me but it's probably not a big deal. I wanted to use `()` as the pong message type, but it serializes to `""` and so the programs in `Node` and `Node.Internal` can't distinguish a sent `()` from an end-of-input (or from two or more sent `()`s). If network-transport preserved the sender's chunking then it would be OK; we'd get a [""] and push that onto the in-channel. But maybe this comes with too high a price (extra data on the wire).